main/main.ino

```diff
@@ -564,10 +577,19 @@ void connectMQTT() {
       Log.warning(F("failed, ssl error code=%d" CR), ((WiFiClientSecure*)eClient)->getLastSSLError());
 #endif
     digitalWrite(LED_ERROR, LED_ERROR_ON);
-    delay(2000);
+    delayWithOTA(2000);
     digitalWrite(LED_ERROR, !LED_ERROR_ON);
-    delay(5000);
+    delayWithOTA(5000);
     if (failure_number_mqtt > maxRetryWatchDog) {
+      unsigned long millis_since_last_ota;
+      while ((millis_since_last_ota != 0) && ((millis_since_last_ota = millis() - last_ota_activity_millis) < ota_timeout_millis)) {
+        // OTA might be still active, we sleep for a while
+        Log.warning(F("OTA might be still active (activity %d ms ago)" CR), millis_since_last_ota);
+#if defined(ESP8266) || defined(ESP32)
+        ArduinoOTA.handle();
+#endif
+        delay(100);
```

Comment:
Are you sure we need a delay() here ?

Reply:
Well, the delay makes the waiting more CPU efficient? There is also that logging on every iteration of the loop currently...

Alternatively we could refactor the code here to log once, and then call delayWithOTA with the "remaining time". Naturally there would be still that delay, inside delayWithOTA.

```

while ((millis_since_last_ota != 0) && ((millis_since_last_ota = millis() - last_ota_activity_millis) < ota_timeout_millis)) {
  Log.warning(F("OTA might be still active (activity %d ms ago)" CR), millis_since_last_ota);
  unsigned long remaining = ota_timeout_millis - millis_since_last_ota;
// TODO: ensure delayWithOTA works properly with non-round numbers
  delayWithOTA(remaining);

}

```

Reply:
One thing I noted, the while clause in the PR was supposed to be

```
while ((last_ota_activity_millis != 0) && ((millis_since_last_ota = millis() - last_ota_activity_millis) < ota_timeout_millis)) {

```

That is, break immediately when last_ota_activity_millis is zero. The zero value signifies no OTA or OTA that has finished.

Reply:
Hmm, do you have a link to an example? We normally would not need any `delay()` whenever `loop()` is concerned. 

However, here I wanted to keep the changes minimal and avoid yielding control to delay. Instead, as we have retried all MQTT already, it's time to reboot (as long as there is no OTA in progress). 

I can remove the delay here and in `delayWithOTA` as well, it should not be too bad as this is an very exceptional code path, should not happen often, and thus CPU effieciency is not everything.

Reply:
I'm just referring to the examples delivered with the [ArduinoOTA library ](https://github.com/espressif/arduino-esp32/tree/master/libraries/ArduinoOTA/examples/BasicOTA)

Reply:
Right yeah, well in that ArduinoOTA library example the logic is in the loop. A bit trickier here IMO.

Configuration I used for tests is here: https://github.com/ssalonen/OpenMQTTGateway/blob/lerbacka/ble_scanner_cont_openhab_env.ini . ESP32 dev board

There's been some slight changes after my test, I could try it one more time to be sure.

Reply:
@1technophile  tested with https://github.com/1technophile/OpenMQTTGateway/pull/1307/commits/6a1b2b04867817d4627b4ceac549d2a0714c26db included [my branch which is rebased on this one](https://github.com/ssalonen/OpenMQTTGateway/blob/lerbacka/ble_scanner_cont_openhab_env.ini), seems to work very reliably.

Reply:
Cool, I will test and merge if ok

Reply:
Working good!